Add/remove networks when registry changes

dcbw opened the pull request:
Add a new -watch-networks option to watch for registry (either
remote or local) changes and add/remove networks for those events.
If this option is given along with -networks, only the networks
specified in -networks will be acted upon.  If -networks is
empty, then all network events will be acted upon.

eyakubovich commented on network/manager.go:
```diff
@@ -0,0 +1,296 @@
+					if _, ok := m.networks[netname]; ok {
+						continue
+					}
+				log.Warningf("Adding network %v", netname)
```
gofmt

dcbw replied:
> gofmt. Also, make these Infos instead of Warnings?

That was actually a left-over from debugging that I've since removed.

eyakubovich commented on network/manager.go:
```diff
@@ -0,0 +1,296 @@
+
+func (m *Manager) RunNetwork(net *Network, wg sync.WaitGroup) {
+	wg.Add(1)
+	defer wg.Done()
```
Stylistic nit: so far I have been following the convention of putting the wait group's add/done calls outside:

```
go func() {
    wg.Add(1)
    defer wg.Done()
    nm.RunNetwork(net)
}
```

I'm not sure if it's "the right way" but would like to keep it for consistency.

eyakubovich replied:
I just updated the comment above to reflect that `wg.Add(1)` should be before starting the goroutine to avoid a race.

eyakubovich commented on network/manager.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+func (m *Manager) watchNetworks(ctx context.Context, wg sync.WaitGroup) {
+	wg.Add(1)
```
Same comment as above.

eyakubovich commented:
@dcbw Apologies for slow review -- away at ContainerCon and such.

dcbw commented:
> @dcbw Apologies for slow review -- away at ContainerCon and such.

No problem; I was on vacation last week anyway.

dcbw commented:
Rebased and repushed with fixes for comments...

eyakubovich commented on network/network.go:
```diff
@@ -26,30 +26,36 @@ import (
 
-func New(sm subnet.Manager, name string, ipMasq bool) *Network {
+func NewNetwork(sm subnet.Manager, name string, ipMasq bool) *Network {
+	ctx, cancel := context.WithCancel(context.Background())
```
I recently realized that a slicker way is to wrap ctx with another `WIthCancel` (https://godoc.org/golang.org/x/net/context#WithCancel). This way you get another `cancel` function but the outer one still works. This way you don't have to:

```
<-ctx.Done()
for _, net := range m.networks {
    net.Cancel()
}
```

dcbw replied:
Oh nice.  Done.

eyakubovich commented on network/manager.go:
```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+func (m *Manager) watchNetworks(ctx context.Context, wg sync.WaitGroup) {
```
passing `WaitGroup` by value actually copies it and I think will cause trouble. Just create a new `WaitGroup` here and wait for it before returning from this function.

dcbw commented:
Rebased on backend-split-init, updated for your comments, and a further fix to the previous getNetworks() functionality to match what etcd actually returns.

eyakubovich commented:
@dcbw Can you rebase again on top of master (just merged your singleton PR). This will make it easier to review.

dcbw commented:
@eyakubovich will do.  I'm working on a few fixups first though that were impacting further testing of this branch.  Hope to get those and another push of this branch done today.

eyakubovich commented on network/manager.go:
```diff
@@ -0,0 +1,306 @@
+			daemon.SdNotify("READY=1")
+		}
+
+				log.Warningf("Running network %v", net.Name)
```
`log.Infof` and also fix indent / run gofmt

eyakubovich commented:
@dcbw LGTM. Will merge after that indent fix.

dcbw commented:
Repushed with fixups; also added network package to 'test' to get gofmt checking.  Should the 'gofmt -l' be changed to 'gofmt -d' to display diffs?  That's been useful to me in the past.
